connector/connect/src/main/protobuf/spark/connect/base.proto

```diff
@@ -100,18 +70,138 @@ message AnalyzePlanRequest {
+  message IsLocal { }
+
+  // Returns true if this Dataset contains one or more sources that continuously
+  // return data as it arrives.
+  message IsStreaming { }
+
+  // Explains the input plan based on a configurable mode.
+  message Explain {
+    // Plan explanation mode.
+    enum ExplainMode {
+      MODE_UNSPECIFIED = 0;
+
+      // Generates only physical plan.
+      SIMPLE = 1;
+
+      // Generates parsed logical plan, analyzed logical plan, optimized logical plan and physical plan.
+      // Parsed Logical plan is a unresolved plan that extracted from the query. Analyzed logical plans
+      // transforms which translates unresolvedAttribute and unresolvedRelation into fully typed objects.
+      // The optimized logical plan transforms through a set of optimization rules, resulting in the
+      // physical plan.
+      EXTENDED = 2;
+
+      // Generates code for the statement, if any and a physical plan.
+      CODEGEN = 3;
+
+      // If plan node statistics are available, generates a logical plan and also the statistics.
+      COST = 4;
+
+      // Generates a physical plan outline and also node details.
+      FORMATTED = 5;
+    }
+
+    // (Required) For analyzePlan rpc calls, configure the mode to explain plan in strings.
+    ExplainMode explain_mode= 1;
+  }
+
+  // Generate a string to express the schema in a nice tree format.
+  // It will invoke 'StructType.treeString' (same as 'Dataset.printSchema')
+  // to compute the results.
+  message TreeString {
+
+    // (Optional) The level to generate the string.
+    optional int32 level = 1;
+  }
+
+  // Returns a best-effort snapshot of the files that compose this Dataset.
+  // It will invoke 'Dataset.inputFiles' to compute the results.
+  message InputFiles { }
+
+  // Returns a `hashCode` of the logical query plan.
+  // It will invoke 'Dataset.semanticHash' to compute the results.
+  message SemanticHash { }
+
+  // Returns `true` when the logical query plans inside both Datasets are equal.
+  // It will invoke 'Dataset.sameSemantics' to compute the results.
+  message SameSemantics {
+    Relation other = 1;
+  }
 }
 
 // Response to performing analysis of the query. Contains relevant metadata to be able to
 // reason about the performance.
 message AnalyzePlanResponse {
   string client_id = 1;
-  DataType schema = 2;
 
-  // The extended explain string as produced by Spark.
-  string explain_string = 3;
+  repeated AnalysisResult results = 2;
+
+  message AnalysisResult {
+    oneof result {
+      // The analyzed schema.
+      DataType schema = 1;
+
+      // Is local
+      bool is_local = 2;
+
+      // Is Streaming
+      bool is_streaming = 3;
+
+      // The extended explain string as produced by Spark.
+      string explain_string = 4;
+
+      // Get the tree string of the schema.
+      string tree_string = 5;
+
+      // Get the input files.
+      InputFiles input_files = 6;
+
+      // Get the semantic hash code.
+      int32 semantic_hash = 7;
```

Comment:
there is no symmetry to the request so it should't be in the request. What is the value of this for the customer? Is this part of the Spark public API?

Do we need this for Spark Connect now?

Reply:
the methods added here are all public API, and used by the users

Reply:
printSchema is frequently used, but I also add others by the way